pennylane/workflow/qnode.py

```diff
@@ -110,15 +110,33 @@ def _to_qfunc_output_type(
     if has_partitioned_shots:
         return tuple(_to_qfunc_output_type(r, qfunc_output, False) for r in results)
 
+    qfunc_output_leaves, qfunc_output_structure = qml.pytrees.flatten(
+        qfunc_output, is_leaf=lambda obj: isinstance(obj, (qml.measurements.MeasurementProcess))
+    )
+    num_of_measurements = len(qfunc_output_leaves)
+
+    results_leaves, results_structure = qml.pytrees.flatten(results)
+    num_of_results = len(results_leaves)
+
     # Special case of single Measurement in a list
     if isinstance(qfunc_output, list) and len(qfunc_output) == 1:
         results = [results]
 
-    # If the return type is not tuple (list or ndarray) (Autograd and TF backprop removed)
-    if isinstance(qfunc_output, (tuple, qml.measurements.MeasurementProcess)):
+    if isinstance(qfunc_output, qml.measurements.MeasurementProcess):
         return results

```

Comment:
I'm also not sure we need this logic anymore, see:
```
qfunc_output = qml.expval(qml.Z(0))

leaves, structure = qml.pytrees.flatten(qfunc_output, lambda obj: isinstance(obj, qml.measurements.MeasurementProcess))

results = 0.2
if len(leaves) == 1: # add dimension back in when measurement dim squeezed out
    results = (results,)
qml.pytrees.unflatten(results, structure)
```
```
0.2
```

`return measurement` should be able to still be handled by the flatten-> unflatten logic.